include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -152,18 +152,77 @@ __pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first
     }
 }
 
+// TODO In C++20 we may try to use std::equality_comparable
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct is_equality_comparable : std::false_type
+{
+};
+
+// Two sycl_iterator's (with different access modes)
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+struct is_equality_comparable<sycl_iterator<_Mode1, _T, _Allocator>, sycl_iterator<_Mode2, _T, _Allocator>>
+    : std::true_type
+{
+};
+
+// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-constexpr bool
+struct is_equality_comparable<_Iterator1, _Iterator2,
+                              std::void_t<decltype(::std::decay_t<_Iterator1>() == ::std::decay_t<_Iterator2>())>>
+    : std::true_type
+{
+};
+
+// Check the ability to compare two sycl_iterator's
+/*
+ * @return bool - true, if these iterators may be the same;
+ *         false - we know these iterators are different
+ */
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+bool
+__check_if_iterator_equality_is_possible(sycl_iterator<_Mode1, _T, _Allocator> __it1,
```

Comment:
Why just `operator==` has not been added to sycl_iterator?

Reply:
Because as said @akukanov, the implementation of sycl_iterator is technically correct. And I agree with this.

Reply:
True, the current implementation of iterators is technically correct. But if we extend it to also check the equality of the buffers, it will also be technically correct, just somewhat slower.

The problem, however, is that such implementation would provide stronger comparison guarantees than those of a typical C++ iterator (and likely also stronger than what we want to document in the specification). Therefore algorithms that work with generic iterators would still not be able to use these guarantees, and would have to first ensure that performing the comparison would be valid.